### PR TITLE
Clippy linting and CI build/test.

### PR DESCRIPTION
**Pull request by Xithrius**
Hello!

I saw your post on the Rust community Discord, and thought I'd contribute this project.

The clippy group lints I've added are nursery and pedantic, which can be found [here](https://rust-lang.github.io/rust-clippy/master/index.html). There are some rules that I ignored that can be seen at the [top of `lib.rs`](https://github.com/Xithrius/rust_search/blob/feature/ci-and-clippy/src/lib.rs#L2-L10).

I wasn't sure how to explain what the [similarity sort function](https://github.com/Xithrius/rust_search/blob/feature/ci-and-clippy/src/utils.rs#L83-L92) panics on, so I added a placeholder for the time being.

The CI builds and tests when there's a commit to main or any PR, and runs on the following operating systems and architectures:

```
windows-latest: x86_64-pc-windows-msvc
ubuntu-latest: x86_64-unknown-linux-gnu
macos-latest: x86_64-apple-darwin
```

Happy holidays!

### CODE REVIEW

**Comment by ParthJadhav** [comment]
Hello there @Xithrius ,

Firstly thanks for making this contribution in a holiday season 😄

Few questions:

1. Why do we need to build it for all the three platforms when it's gonna directly be downloaded from the cargo. Is it that, If I compile or publish from MacOS the other platforms won't be able to use the crate? Also it doesn't build on ARM for MacOS would that cause any issues for users on M Series Macs ?

2. The Build command in the workflow... does it need to have -r for the release version or it's okay not to have it.

3. The lint and formatting runs after the build? Shouldn't it run before so we catch certain ill practices before hand?

I'm new to Rust development. Please excuse dumb questions.

**Comment by Xithrius** [comment]
1. Building for all 3 platforms is just to make sure that there's no broken or extra dependencies needed. For instance, one of my applications broke when open-ssh needed something more for macos. If you build for all 3 and it works, publishing from any of them will work for the rest.
2. It's ok not to have it.
3. If there's a syntax error that the build command catches or something like that, then it won't be that useful to run the lint in my opinion.

**Comment by ParthJadhav** [comment]
Ohh, Intresting @Xithrius . Got it. I'll test it once and we can merge it.

**Comment by ParthJadhav** [comment]
One more thing @Xithrius , The builds for different platforms... How are they gonna be used ? I mean

Would that have any effect on the `cargo publish` command. or is it just to test if everything works properly?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,83 @@
+name: CI
+
+on:
+  push:
+    branches: [ main ]
```
Comment (ParthJadhav) [inline .github/workflows/ci.yml]: This should be `master` right?

**Review by ParthJadhav — APPROVED** [review]
Just left few comments

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Xithrius) [reply .github/workflows/ci.yml]: Indeed.

Comment (Xithrius) [reply .github/workflows/ci.yml]: Resolved in [d30a47a](https://github.com/ParthJadhav/rust_search/pull/30/commits/d30a47a1c6db98b981c4a49e17604c3acc259fa5).

**Comment by Xithrius** [comment]
> One more thing @Xithrius , The builds for different platforms... How are they gonna be used ? I mean
> 
> Would that have any effect on the `cargo publish` command. or is it just to test if everything works properly?

@ParthJadhav Correct, it is only for testing. `cargo publish` won't be effected.

**Comment by ParthJadhav** [comment]
Got it @Xithrius , Will merge it 🚀 Thanks for the enhancements.

**Pull request merged**